astroplan/scheduling.py

```diff
@@ -140,6 +140,76 @@ def create_score_array(self, time_resolution=1*u.minute):
             score_array *= constraint(self.observer, targets, times)
         return score_array
 
+    def weighted_score_array(self, time_resolution=1*u.minute):
+        """
+        For each block, this returns a score for each time using the
+        formula (score*weight+score*weight+...)/(weight+weight+..)
+        """
+        # note: with this method, a non-boolean constraint with weight
+        # zero has the exact same effect if it was boolean
+        start = self.schedule.start_time
+        end = self.schedule.end_time
+        times = time_grid_from_range((start, end), time_resolution)
+        score_array = np.zeros((len(self.blocks), len(times)))
+        zeros = np.ones((len(self.blocks), len(times)), dtype=int)
+        local_constraints = []
+        weights = []
+        for i, block in enumerate(self.blocks):
+            weights.append(0)
+            local_constraints.append([])
+            # schedulers put global constraints into ._all_constraints
+            # so we can use .constraints for the local constraints
+            if block.constraints:
+                for constraint in block.constraints:
+                    local_constraints[i].append(constraint.__class__.__name__)
+                    applied_score = constraint(self.observer, [block.target],
+                                               times=times)[0]
+                    if constraint.boolean_constraint:
+                        # add to the mask designating where the score is zero
+                        zeros[i] &= applied_score
+                    # TODO: make a default weight=1 and merge these
+                    elif constraint.weight:
+                        zeros[i][np.where(applied_score == 0)] = 0
+                        weight = constraint.weight
+                        weights[i] += weight
+                        score_array[i] += applied_score * weight
+                    else:
+                        zeros[i][np.where(applied_score == 0)] = 0
+                        weights[i] += 1
+                        score_array[i] += applied_score
+        targets = [block.target for block in self.blocks]
+        for constraint in self.global_constraints:
+            skip_global = []
+            for i, block in enumerate(local_constraints):
+                if constraint.__class__.__name__ in block:
+                    skip_global.append(i)
+            global_score = constraint(self.observer, targets, times)
+            if constraint.boolean_constraint:
+                zeros &= 1-global_score
```

Comment:
I have no idea why the `1-` is there (might have been a relic from a previous zero-tracking method). It is fixed in the next commit.